src/kernel/filesystem/fat32.zig

```diff
@@ -652,6 +659,445 @@ pub fn Fat32FS(comptime StreamType: type) type {
+            ///     Fat32Self.Error - Error reading the cluster chain if the buffer isn't aligned.
+            ///     ReadError       - Error reading from the underlying stream.
+            ///     SeekError       - Error seeking the underlying stream.
+            ///
+            pub fn next(self: *EntryIteratorSelf) (Allocator.Error || Fat32Self.Error || ReadError || SeekError || LongName.Error)!?Entry {
+                // If there is a orphan file, then just get the next one
+                // If we hit the end of the cluster chain, return null
+                while (true) {
+                    if (self.nextImp()) |n| {
+                        return n;
+                    } else |e| switch (e) {
+                        error.Orphan => continue,
+                        error.EndClusterChain => return null,
+                        else => return @errSetCast(Allocator.Error || Fat32Self.Error || ReadError || SeekError || LongName.Error, e),
+                    }
+                }
+            }
+
+            ///
+            /// Destroy the entry iterator freeing any memory.
+            ///
+            /// Arguments:
+            ///     IN self: *EntryIteratorSelf - Iterator self.
+            ///
+            pub fn deinit(self: *const EntryIteratorSelf) void {
+                self.allocator.free(self.cluster_block);
+                self.cluster_chain.deinit();
+            }
+
+            ///
+            /// Initialise a directory entry iterator to looping over the FAT entries. This uses
+            /// the cluster chain iterator.
+            ///
+            /// Arguments:
+            ///     IN allocator: *Allocator - The allocator for allocating a cluster block cache.
+            ///     IN fat_config: FATConfig - The FAT configuration.
+            ///     IN cluster: u32          - The first cluster to start reading from.
+            ///     IN stream: StreamType    - The underlying stream.
+            ///
+            /// Return: EntryIteratorSelf
+            ///     The entry iterator.
+            ///
+            /// Error: Allocator.Error || Fat32Self.Error || ReadError || SeekError
+            ///     Allocator.Error - Error allocating memory for fields in the return entry.
+            ///     Fat32Self.Error - Error reading the cluster chain if the buffer isn't aligned.
+            ///     ReadError       - Error reading from the underlying stream.
+            ///     SeekError       - Error seeking the underlying stream.
+            ///
+            pub fn init(allocator: *Allocator, fat_config: FATConfig, cluster: u32, stream: StreamType) (Allocator.Error || Fat32Self.Error || ReadError || SeekError)!EntryIteratorSelf {
+                var cluster_block = try allocator.alloc(u8, fat_config.bytes_per_sector * fat_config.sectors_per_cluster);
+                errdefer allocator.free(cluster_block);
+                var it = try ClusterChainIterator.init(allocator, fat_config, cluster, stream);
```

Comment:
Does this need an errdefer to clean up?

Reply:
it does, and forgot to add tests for this init function